packages/core/src/PartialExecutionUtils/recreateNodeExecutionStack.ts

```diff
@@ -0,0 +1,166 @@
+import {
+	NodeConnectionType,
+	type IExecuteData,
+	type INode,
+	type INodeExecutionData,
+	type IPinData,
+	type IRunData,
+	type ISourceData,
+	type ITaskDataConnectionsSource,
+	type IWaitingForExecution,
+	type IWaitingForExecutionSource,
+} from 'n8n-workflow';
+
+import * as a from 'assert/strict';
+import type { DirectedGraph } from './DirectedGraph';
+import { getIncomingData } from './getIncomingData';
+import { getSourceDataGroups } from './getSourceDataGroups';
+
+export function recreateNodeExecutionStack(
+	graph: DirectedGraph,
+	startNodes: INode[],
+	destinationNode: INode,
+	runData: IRunData,
+	pinData: IPinData,
+): {
+	nodeExecutionStack: IExecuteData[];
+	waitingExecution: IWaitingForExecution;
+	waitingExecutionSource: IWaitingForExecutionSource;
+} {
+	// Validate invariants.
+
+	// The graph needs to be free of disabled nodes. If it's not it hasn't been
+	// passed through findSubgraph2.
+	for (const node of graph.getNodes().values()) {
+		a.notEqual(
+			node.disabled,
+			true,
+			`Graph contains disabled nodes. This is not supported. Make sure to pass the graph through "findSubgraph2" before calling "recreateNodeExecutionStack". The node in question is "${node.name}"`,
+		);
+	}
+
+	// Initialize the nodeExecutionStack and waitingExecution with
+	// the data from runData
+	const nodeExecutionStack: IExecuteData[] = [];
+	const waitingExecution: IWaitingForExecution = {};
+	const waitingExecutionSource: IWaitingForExecutionSource = {};
+
+	// TODO: Don't hard code this!
+	const runIndex = 0;
+
+	for (const startNode of startNodes) {
+		const incomingStartNodeConnections = graph
+			.getDirectParents(startNode)
+			.filter((c) => c.type === NodeConnectionType.Main);
```

Comment:
This could be extracted into `isMainConnection` or something with a better name

Reply:
Added to the list of todos in https://github.com/n8n-io/n8n/pull/10758/files#r1752298139